Support small screens

TiagoCavalcanteTrindade opened the pull request:
This branch is made on the top of #67.

This branch add small screen support for graphics 3d.

This branch updates the point size.

![size](https://user-images.githubusercontent.com/62714153/121541762-f4fb5700-c9dd-11eb-9e5b-8193fd2ec52d.gif)

TiagoCavalcanteTrindade commented on mathics_django/web/format.py:
```diff
@@ -65,21 +79,16 @@ def format_output(obj, expr, format=None):
             result = expr.format(obj, "System`OutputForm")
         elif head == 'System`String':
             result = expr.format(obj, "System`InputForm")
+        elif head == 'System`Graphics3D':
```
@rocky this won't, it depends in the value 400 be passed to the function `drawGraphics3d` (what MathML won't do).

To this works the parameter need be replaced by a constant.

TiagoCavalcanteTrindade replied:
@rocky I replaced the parameter with a constant and the axes number bugged, then I remembered, the absolute size is relative to the container, not the canvas size. Fixing it now.

TiagoCavalcanteTrindade replied:
@rocky I hadn't success, but I guess it'll work when MathML be removed of graphics 3d.

rocky replied:
Here is something save and will reduce the PR changes: reformat javascript according to the style mentioned in editorconfig. 


Which reminds me. Editorconfig should be changed so that it just applies to Javascript.  Python files should stay as they are and those are formatted by Python black.

TiagoCavalcanteTrindade replied:
@rocky I'm formatting the JavaScript and CSS now.

Could you confirm the Python rules to me?
- indent style: spaces
- indent size: 4
- insert final newline: true

rocky replied:
Also, I guess we'll have to note skipping those css and js files such as for mathjax, and threejs that are not ours but are copied from elsewhere into our space.

TiagoCavalcanteTrindade replied:
@rocky I'm doing it.

Also should the `three.js` file be deleted (we are using `three.min.js`)?

rocky replied:
@TiagoCavalcanteTrindade  : 
Install optional dependency used in Mathics Core [psutil](https://pypi.org/project/psutil/). We should also fix the python code so that it doesn't error out if psutil  or SystemMemory isn't available.

TiagoCavalcanteTrindade commented:
@rocky ok.

There are lots of comments which I think are unnecessary in `base_html.html`. Could I delete them?

rocky commented:
> @rocky ok.
> 
> There are lots of comments which I think are unnecessary in `base_html.html`. Could I delete them?

Sure, but let's put them it its own PR. Many small PRs is good.

TiagoCavalcanteTrindade commented:
@rocky I'll do it.

Also could I add [psutil](https://pypi.org/project/psutil/) as an optional dependency?

rocky commented:
> @rocky I'll do it.

Thanks. 
> 
> Also could I add [psutil](https://pypi.org/project/psutil/) as an optional dependency?

It doesn't make sense to me in that this is a dependency of Mathics core, not Mathics Django. Mathics Django should only be concerned about whether there is SystemMemory. (Maybe one day we'll figure that out some other way). 

When each package lists explicitly all the transitive dependencies it causes a mess.

BTW, over the weekend I will be updating the release version in Mathics Core to something like 3.0.0.dev0 because of the API breakage (and I plan to piggyback more breaking API changes in this that would be nice).

And then in the these branches that depend on that API change they will up their minimum dependency too.

TiagoCavalcanteTrindade commented:
@rocky this makes all sense.

> BTW, over the weekend I will be updating the release version in Mathics Core to something like 3.0.0.dev0 because of the API breakage (and I plan to piggyback more breaking API changes in this that would be nice).
> And then in the these branches that depend on that API change they will up their minimum dependency too.

Ok.

TiagoCavalcanteTrindade commented:
@rocky I removed the bloat and fixed it.

You don't think 200 of multiplier is too small?
![gif](https://user-images.githubusercontent.com/62714153/121721649-dc14a380-caba-11eb-9141-a103bf8639fc.gif)
(boring MathML)

Here is 400:
![image](https://user-images.githubusercontent.com/62714153/121721723-efc00a00-caba-11eb-86e7-407e3c51638b.png)

TiagoCavalcanteTrindade reviewed (commented):
@rocky for me removing this line still [looks like to] have the same behavior.

rocky commented:
> @rocky for me removing the last line of the file `mathics.js` still [looks like to] have the right behavior.

@TiagoCavalcanteTrindade  Ok. And I don't think 200 is too small. When CI clears just merge this.

TiagoCavalcanteTrindade commented:
@rocky I'll set it to 200.

And remove the line I told you too.

rocky commented:
> @rocky I'll set it to 200.
> 

Another great reason for tolerating smaller sizes is that in the future nesting inside a list like we see in the 3D Graphics Test Suite: 

```
Table[Graphics3D[{PointSize[r],```` Point[{0, 0, 0}]}], {r, {0.05, 0.1, 0.8}}]
```

Nesting inside a list like this will force the inner 3D graphics to a smaller size.

TiagoCavalcanteTrindade commented:
@rocky this makes sense.

Removing Prototype and Scryptaculous has being easier than I thought it would be (also more boring 😄).

When I finish I'll improve the polygon drawing (using earcut).

rocky commented:
> @rocky this makes sense.
> 
> Removing Prototype and Scryptaculous has being easier than I thought it would be (also more boring ).
> 
> When I finish I'll improve the polygon drawing (using earcut).

I've been thinking on and off about this.  The big picture is this...

Mathics Core is supposed to be a computational engine that can do just about any calculation you give it. 
In the past (or I suppose currently) it makes too many calculations for graphics. By too many I mean it calculates to the extent that the front-end render is too constrained to be able adjust flexibly. For example if I resize the window and everything is calculated for me, then I can't easily resize. In the past Herculean and Heroic attempts have been made to try to undo this, but this is the wrong idea.

So now we are giving more power to the front end to decide what to do. And the front end can, when it wants, call Mathics Core as many times as it wants to compute more stuff it needs.

What this has to do here, is that the front end in checking whether the point are coplanar could also report back what the plane is. I suppose it could also detect or describe where the holes are. 

SVG in 2D seems to be able to figure things out. So the algorithm it uses could be used if we pass back what the plain is and map this from 3D into 2D orienting the plain before rendering.

TiagoCavalcanteTrindade commented:
@rocky this is very complicated.

If something is computed on all front ends it should be computed in the backend. But graphics 3d isn't used in all front ends (e.g.: mathicsscript).

Also Mathics Core is [or at least I guess should be] a core before a backend, and good features of a core is be small and have no bloat, and if I'm using mathicsscript 3d calculus is bloat.

Also calculate the extent is front end dependent (a front end can show just part of the graphic and another can have the camera at 1 mile of distance of the scene).

---

About the polygon I thought in the following:
if triangle draw triangle
else if coplanar calculate holes and draw the path with [holes](https://threejs.org/docs/#api/en/extras/core/Shape.holes)
else earcut the polygon and draw it

mmatera commented:
A few ideas about this problem which we had pass over before:
* If Graphics3D boxes are not shown in certain front ends, it does not mean that you do not want to be able to generate them: for example, in IWolfram, I use the terminal front-end to produce encoded output that then is reader by the jupyter kernel and converted to something showable. Also, it is useful to implement "Export" functions.
* In order to avoid putting too much stuff in the core, one idea was to implement special symbols and functionalities as "Pymathics" modules or also as WL packages, that the front-end can load if it uses it. One example of that is the projects pymathics-matplotlib and pymathics-asymptote. Then, the front-end can just load the module that it uses.
* One idea to implement the feedback you want between core and front end could be just to set WL variables when you need it from the front end. Then, if your front-end changes the size of the viewport, it can set the size of the graphics as a variable in a private context of the WL session, and then ask to rebuild the plot.

TiagoCavalcanteTrindade commented:
> - If Graphics3D boxes are not shown in certain front ends, it does not mean that you do not want to be able to generate them: for example, in IWolfram, I use the terminal front-end to produce encoded output that then is reader by the jupyter kernel and converted to something showable. Also, it is useful to implement "Export" functions.

I didn't think like this, but now makes sense.

> - In order to avoid putting too much stuff in the core, one idea was to implement special symbols and functionalities as "Pymathics" modules or also as WL packages, that the front-end can load if it uses it. One example of that is the projects pymathics-matplotlib and pymathics-asymptote. Then, the front-end can just load the module that it uses.

@mmatera this is a great idea!  
Also, I was thinking if the packages DiscreteMath and VectorAnalysis should be moved to their own repository, what are your and @rocky thoughts?

rocky commented:
> > * If Graphics3D boxes are not shown in certain front ends, it does not mean that you do not want to be able to generate them: for example, in IWolfram, I use the terminal front-end to produce encoded output that then is reader by the jupyter kernel and converted to something showable. Also, it is useful to implement "Export" functions.

We've gone over "Export" functions before. Yes, they are good, but the way it is implemented right now with taking substrings instead of using structured object is hacky. We need to fix up the Boxing primitives too.  

But I totally understand that we can't fix up everything all at once, so `if` it were the case that IWolfram needed it right now and were it not for the export functions getting fixed up, then sure, let's exend the hacky approach to more things. 

However IWolfram is not close to working with all the recent changes to Mathics, and was doing things in a somewhat hacky way itself. The longer we can wait on this the better off we will be. When IWolfram is revised it can be done in a simpler, and more direct way.


> 
> I didn't think like this, but now makes sense.
> 
> > * In order to avoid putting too much stuff in the core, one idea was to implement special symbols and functionalities as "Pymathics" modules or also as WL packages, that the front-end can load if it uses it. One example of that is the projects pymathics-matplotlib and pymathics-asymptote. Then, the front-end can just load the module that it uses.

We discussed this a little bit too. Absolutely. The agreement at least for now was to have a version of svg and asymptote (and some JSON/js) in there right now because 

* we don't have the Pymathics modules yet
* we probably need to provide minimal _some_ level of support for something without any Pymathics modules. The support can be kind of crappy as is the case when you compare the mathics CLI versus mathicsscript which does completion and fancy code colorization
*  these routines don't add any external dependencies (e.g. on some svg or JSON library)
* Pymathics modules still need a little bit of work in terms of how they handle documention

> 
> @mmatera this is a great idea!
> 
> Also, I was thinking if the packages DiscreteMath and VectorAnalysis should be moved to their own repository, what are your and @rocky thoughts?

I alluded to some of this above. The things I use to guide this are: 

* Are there external libraries needed to support these? For example NaturalLanguage processing was moved out because it relies on a somewhat heavy library (NLTK) which also pulls in a bit of stuff.
* Is this a core part of WL. If yes, then it is better to include it in Core except where there is the concern above

In the case of both of these packages, I think neither is requires extra libraries and both are expected to be in WL. Therefore I think it better for these to remain. Otherwise there will be a burden on users to add these and load them that is not normally needed in Mathematica.

TiagoCavalcanteTrindade commented:
@rocky as they don't have any dependencies I think make sense keep them.
